### PR TITLE
PI 68: Current and Historic Order Supervisor on Community API Convictions

### PR DESCRIPTION
**Pull request by trozzomoj**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/jpa/standard/entity/Enforcement.java [hunk 1]
```diff
@@ -36,6 +37,7 @@
 @Entity
 @Table(name = "ENFORCEMENT")
 @Where(clause = "SOFT_DELETED = 0")
+@EqualsAndHashCode(callSuper=false)
```
Comment (mxco86) [inline src/main/java/uk/gov/justice/digital/delius/jpa/standard/entity/Enforcement.java:40]: Is this change related to the work?

File: src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java [hunk 2]
```diff
@@ -81,27 +85,28 @@ public ConvictionTransformer() {
...
+                .orElse(null))
+            .inBreach(event.isInBreach())
+            .failureToComplyCount(event.getFailureToComplyCount())
+            .breachEnd(event.getBreachEnd())
+            .latestCourtAppearanceOutcome(courtAppearances.map(ConvictionTransformer::latestOutcomeOf).orElse(null))
+            .responsibleCourt(Optional.ofNullable(event.getCourt()).map(CourtTransformer::courtOf).orElse(null))
+            .courtAppearance(courtAppearances.map(CourtAppearanceBasicTransformer::latestOrSentencingCourtAppearanceOf).orElse(null))
+            .awaitingPsr(isAwaitingPsr(event.getDisposal(), courtAppearances))
+            .build();
     }
```
Comment (mxco86) [inline src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java:110]: I think the whole file has had it's indentation reformatted - can we revert the whitespace changes so we can see the relevant code updates?

File: src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.Optional;
+
+public class OrderManagerTransformer {
+
+    public static OrderManager orderManagerOf(uk.gov.justice.digital.delius.jpa.standard.entity.OrderManager entity){
+        return OrderManager.builder().
+            name(Optional.ofNullable(entity.getStaff().getForename()).orElse("").
+                concat(" "+Optional.ofNullable(entity.getStaff().getSurname()).orElse(""))).
+            staffCode(Optional.ofNullable(entity.getStaff().getStaffId()).map(x->x.toString()).orElse("")).
```
Comment (mxco86) [inline src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java:13]: Can we name the lambda argument `staffId` rather than `x`?

File: src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java (hunk 2 above)
Comment (trozzomoj) [reply src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java:110]: no worries.

File: src/main/java/uk/gov/justice/digital/delius/jpa/standard/entity/Enforcement.java (hunk 1 above)
Comment (trozzomoj) [reply src/main/java/uk/gov/justice/digital/delius/jpa/standard/entity/Enforcement.java]: we need to look into , definitely need to be changed

File: src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java [hunk 4]
```diff
@@ -142,6 +147,12 @@ private static KeyValue outcomeOf(StandardReference outcome) {
                 .build();
     }
 
+    private static List<OrderManager> orderManagersOf(@NotNull Event event){
+        return Optional.ofNullable(event.getOrderManagers()).map(x->x.stream().map(OrderManagerTransformer::orderManagerOf).collect(Collectors.toList())).orElse(Collections.emptyList());
```
Comment (mxco86) [inline src/main/java/uk/gov/justice/digital/delius/transformers/ConvictionTransformer.java:151]: Can we name the lambda argument `orderManager`

File: src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import uk.gov.justice.digital.delius.data.api.OrderManager;
+
+import java.util.Optional;
+
+public class OrderManagerTransformer {
+
+    public static OrderManager orderManagerOf(uk.gov.justice.digital.delius.jpa.standard.entity.OrderManager entity){
+        return OrderManager.builder().
+            name(Optional.ofNullable(entity.getStaff().getForename()).orElse("").
+                concat(" "+Optional.ofNullable(entity.getStaff().getSurname()).orElse(""))).
```
Comment (marcus-bcl) [inline src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java:12]: If the surname is empty here, then we'll end up with an extra space at the end of the name.  Maybe we should return the forenames and surname separately to keep it simple?

Also there is a `forename2` field on the `Staff` entity (like middle name), so if you stick with joining the names then that should be included too. Maybe something like this would be better?
```suggestion
            name(Stream.of(entity.getStaff().getForename(), entity.getStaff().getForename2(), entity.getStaff().getSurname())
				.filter(StringUtils::hasLength)
				.collect(joining(delimiter))).
```
(which would match the method used in the user management tool: https://github.com/ministryofjustice/ndelius-um/blob/main/src/main/java/uk/co/bconline/ndelius/util/NameUtils.java)

File: src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+            staffCode(Optional.ofNullable(entity.getStaff().getStaffId()).map(staffId->staffId.toString()).orElse("")).
+            dateStartOfAllocation(entity.getTeam().getStartDate()).
+            dateEndOfAllocation(entity.getTeam().getEndDate()).
+            officerId(entity.getOrderManagerId()).
+            probationAreaId(entity.getProbationArea().getProbationAreaId()).
+            teamId(entity.getTeam().getTeamId()).build();
```
Comment (marcus-bcl) [inline src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java:18]: Minor thing - any reason for putting the `.` at the end of the line here?  Other places in the project seems to put it at the start.

Comment (trozzomoj) [reply src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java:18]: ok not an issue , I think is just agreed style. I ll put  them on the other side.

File: src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java (hunk 5 above)
Comment (trozzomoj) [reply src/main/java/uk/gov/justice/digital/delius/transformers/OrderManagerTransformer.java:12]: The requirement is just for name. I think should never be null, they are both defined as not null in the database. I have used Optional.ofNUllable just for test purpose, because in 58 tests  are passed as null. The delimiter idea is good and I guess should be an empty space . I think Name utils needs to be reviewed ;) eheh.

**Review by marcus-bcl — APPROVED** [review]
